opm/simulators/linalg/PressureTransferPolicy.hpp

```diff
@@ -0,0 +1,158 @@
+{
+public:
+    typedef Dune::Amg::LevelTransferPolicyCpr<FineOperator, CoarseOperator> ParentType;
+    typedef Communication ParallelInformation;
+    typedef typename FineOperator::domain_type FineVectorType;
+
+public:
+    PressureTransferPolicy(const Communication& comm, const FineVectorType& weights, int pressure_var_index)
+        : communication_(&const_cast<Communication&>(comm))
+        , weights_(weights)
+        , pressure_var_index_(pressure_var_index)
+    {
+    }
+
+    virtual void createCoarseLevelSystem(const FineOperator& fineOperator) override
+    {
+        using CoarseMatrix = typename CoarseOperator::matrix_type;
+        const auto& fineLevelMatrix = fineOperator.getmat();
+        coarseLevelMatrix_.reset(new CoarseMatrix(fineLevelMatrix.N(), fineLevelMatrix.M(), CoarseMatrix::row_wise));
+        auto createIter = coarseLevelMatrix_->createbegin();
+
+        for (const auto& row : fineLevelMatrix) {
+            for (auto col = row.begin(), cend = row.end(); col != cend; ++col) {
+                createIter.insert(col.index());
+            }
+            ++createIter;
+        }
+
+        calculateCoarseEntries(fineOperator);
+        coarseLevelCommunication_.reset(communication_, [](Communication*) {});
+
+        this->lhs_.resize(this->coarseLevelMatrix_->M());
+        this->rhs_.resize(this->coarseLevelMatrix_->N());
+        using OperatorArgs = typename Dune::Amg::ConstructionTraits<CoarseOperator>::Arguments;
+        OperatorArgs oargs(*coarseLevelMatrix_, *coarseLevelCommunication_);
+        this->operator_.reset(Dune::Amg::ConstructionTraits<CoarseOperator>::construct(oargs));
+    }
+
+    virtual void calculateCoarseEntries(const FineOperator& fineOperator) override
+    {
+        const auto& fineMatrix = fineOperator.getmat();
+        *coarseLevelMatrix_ = 0;
+        auto rowCoarse = coarseLevelMatrix_->begin();
+        for (auto row = fineMatrix.begin(), rowEnd = fineMatrix.end(); row != rowEnd; ++row, ++rowCoarse) {
+            assert(row.index() == rowCoarse.index());
+            auto entryCoarse = rowCoarse->begin();
+            for (auto entry = row->begin(), entryEnd = row->end(); entry != entryEnd; ++entry, ++entryCoarse) {
+                assert(entry.index() == entryCoarse.index());
+                double matrix_el = 0;
+                if (transpose) {
+                    auto bw = weights_[entry.index()];
+                    for (size_t i = 0; i < bw.size(); ++i) {
+                        matrix_el += (*entry)[pressure_var_index_][i] * bw[i];
+                    }
+                } else {
+                    auto bw = weights_[row.index()];
+                    for (size_t i = 0; i < bw.size(); ++i) {
+                        matrix_el += (*entry)[i][pressure_var_index_] * bw[i];
+                    }
+                }
+                (*entryCoarse) = matrix_el;
+            }
+        }
+        assert(rowCoarse == coarseLevelMatrix_->end());
```

Comment:
I would have assumed that in a parallel run one would have to modify the ghost rows to represent Dirichlet boundary conditions (one on the diagonal, rest zero). Am I missing something?

Reply:
This code is refactored from createCoarseLevelSystem() in BlackoilAmg. The main difference (aside from splitting in two functions) is that this version deals with an unmodified fine matrix and calculates the weighted coarse entries on the fly, whereas the BlackoiAmg variant works on a fine matrix that has been modified to take the weights into account and just extracts the coarse entry from one of the block entries. The resulting coarse matrices should be the same.

Given that the fine matrix has zero-d out the row entries except for the diagonal which is 1e100, the course matrix should have the same property. Is this the correct coarse matrix for AMG? If it is wrong here, it is also wrong in BlackoilAmg.hpp we think.

Reply:
You are right of course concerning all transfers. I was somehow assuming that there is aggregation but of course there is not. Just one question: How do the weights look in the overlap/ghost region?

Reply:
For the *matrix*, we think it is ok:
 - the weights will be (0, 1, 0) [for blackoil with pressure as index 1] in overlap cells (not owned)
 - this is not the same as the weights will be on the process that owns that cell, but the resulting coarse matrix will end up with the same feature as the fine matrix: zero off-diagonals for overlap rows and 1e100*identity on overlap diagonal.
 - so anything that is good/necessary about that matrix structure on the fine matrix should be propagated correctly to the coarse matrix.

For the *vectors*, we think the moveTo{Coarse/Fine}Level is fine, since it is consistent with the matrix treatment, using the same weights. Any un-communicated values in the vector should end up divided by 1e100.

Related to this: looking at the function `DirichletBoundarySetter<T>::set` we wonder how this is used, and perhaps we should use this or something similar instead of the `makeOverlapRowsInvalid` approach currently used (setting 1e100 on the overlap diagonal etc.).